okhttp/src/main/java/okhttp3/internal/http2/Http2Reader.kt

```diff
@@ -0,0 +1,506 @@
+      if (logger.isLoggable(FINE)) logger.fine(format("<< CONNECTION %s", connectionPreface.hex()))
+      if (CONNECTION_PREFACE != connectionPreface) {
+        throw IOException("Expected a connection header but was ${connectionPreface.utf8()}")
+      }
+    }
+  }
+
+  @Throws(IOException::class)
+  fun nextFrame(requireSettings: Boolean, handler: Handler): Boolean {
+    try {
+      source.require(9) // Frame header size.
+    } catch (e: EOFException) {
+      return false // This might be a normal socket close.
+    }
+
+    //  0                   1                   2                   3
+    //  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+    // +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    // |                 Length (24)                   |
+    // +---------------+---------------+---------------+
+    // |   Type (8)    |   Flags (8)   |
+    // +-+-+-----------+---------------+-------------------------------+
+    // |R|                 Stream Identifier (31)                      |
+    // +=+=============================================================+
+    // |                   Frame Payload (0...)                      ...
+    // +---------------------------------------------------------------+
+    val length = source.readMedium()
+    if (length < 0 || length > INITIAL_MAX_FRAME_SIZE) {
+      throw IOException("FRAME_SIZE_ERROR: $length")
+    }
+    val type = source.readByte() and 0xff
+    if (requireSettings && type != TYPE_SETTINGS) {
+      throw IOException("Expected a SETTINGS frame but was $type")
+    }
+    val flags = source.readByte() and 0xff
+    val streamId = source.readInt() and 0x7fffffff // Ignore reserved bit.
+    if (logger.isLoggable(FINE)) logger.fine(frameLog(true, streamId, length, type, flags))
+
+    when (type) {
+      TYPE_DATA -> readData(handler, length, flags, streamId)
+      TYPE_HEADERS -> readHeaders(handler, length, flags, streamId)
+      TYPE_PRIORITY -> readPriority(handler, length, flags, streamId)
+      TYPE_RST_STREAM -> readRstStream(handler, length, flags, streamId)
+      TYPE_SETTINGS -> readSettings(handler, length, flags, streamId)
+      TYPE_PUSH_PROMISE -> readPushPromise(handler, length, flags, streamId)
+      TYPE_PING -> readPing(handler, length, flags, streamId)
+      TYPE_GOAWAY -> readGoAway(handler, length, flags, streamId)
+      TYPE_WINDOW_UPDATE -> readWindowUpdate(handler, length, flags, streamId)
+      else -> source.skip(length.toLong()) // Implementations MUST discard frames of unknown types.
+    }
+
+    return true
+  }
+
+  @Throws(IOException::class)
+  private fun readHeaders(handler: Handler, length: Int, flags: Int, streamId: Int) {
+    if (streamId == 0) throw IOException("PROTOCOL_ERROR: TYPE_HEADERS streamId == 0")
+
+    val endStream = (flags and FLAG_END_STREAM) != 0
+    val padding = if (flags and FLAG_PADDED != 0) source.readByte() and 0xff else 0
```

Comment:
Is the conditional's precedence correct here? Above you parenthesize the `and` operation before the not-equals check.

Turns out there's a lot of unparenthesized versions below as well, so I assume yes.

Reply:
No action taken!